### PR TITLE
(towards 1134) Move OpenCL set_args subroutine creation into the transformation

### PR DESCRIPTION
**Pull request by sergisiso**
I split this issue once again and did another "towards" PR. In this case to move the OpenCL set_args subroutines into the transformation. As part of this I needed to make the Call more flexible to be found inside sub-expressions.

### CODE REVIEW

File: src/psyclone/domain/gocean/transformations/gocean_opencl_trans.py [hunk 1]
```diff
@@ -250,6 +256,121 @@ def apply(self, node, options=None):
...
+        for arg in kernel.arguments.args:
+
+            name = argsetter.symbol_table.next_available_name(arg.name)
+
+            # This function requires 'TARGET' annotated declarations which are
+            # not supported in the PSyIR, so we build them as UnknownFortraType
+            # for now
+            if arg.is_scalar and arg.intrinsic_type == "real":
+                pointer_type = UnknownFortranType(
+                    "REAL(KIND=go_wp), INTENT(IN), TARGET :: " + name)
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/domain/gocean/transformations/gocean_opencl_trans.py:331]: Nope, that's fine.

File: src/psyclone/domain/gocean/transformations/gocean_opencl_trans.py [hunk 2]
```diff
@@ -250,6 +256,121 @@ def apply(self, node, options=None):
...
+            elif arg.is_scalar:
+                pointer_type = UnknownFortranType(
+                    "INTEGER, INTENT(IN), TARGET :: " + name)
+            else:
+                # Everything else is a cl_mem pointer (c_intptr_t)
+                pointer_type = UnknownFortranType(
+                    "INTEGER(KIND=c_intptr_t), INTENT(IN), TARGET :: " + name)
+
+            new_arg = DataSymbol(name, datatype=pointer_type,
+                                 interface=r_interface)
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/domain/gocean/transformations/gocean_opencl_trans.py]: Great.

File: src/psyclone/psyir/backend/fortran.py [hunk 3]
```diff
@@ -1488,5 +1488,9 @@ def call_node(self, node):
         for child in node.children:
             result_list.append(self._visit(child))
         args = ", ".join(result_list)
-        return "{0}call {1}({2})\n".format(self._nindent, node.routine.name,
-                                           args)
+        if not node.parent or isinstance(node.parent, Schedule):
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/backend/fortran.py:1491]: Oh yes, I should have realised that. Ignore me.

File: src/psyclone/domain/gocean/transformations/gocean_opencl_trans.py [hunk 4]
```diff
@@ -309,11 +309,11 @@ def _insert_ocl_arg_setter_routine(node, kernel):
         argsetter.symbol_table.add(check_status)
 
         # All arguments are read-only, create the interface for later use
```
Comment (arporter) [inline src/psyclone/domain/gocean/transformations/gocean_opencl_trans.py:311]: Please could you remove this comment as it's no longer relevant.

**Review by arporter — CHANGES_REQUESTED** [review]
Thanks Sergi.
There's just one old comment left over that needs removing.
I've brought up-to-date with master.
All tests are OK with compilation.
However, I get a failure for gocean/eg3 when compilation is enabled:

    gfortran  -I../../../external/dl_esm_inf/finite_difference/src -I../../../external/FortCL/src -c psy_gen.f90
    psy_gen.f90:72:23:

       72 | &ystart - 1, ystop - 1)
          |                       1
    Error: More actual than formal arguments in procedure call at (1)
    psy_gen.f90:87:40:

       87 | &xstop_1 - 1, ystart_1 - 1, ystop_1 - 1)
          |                                        1
    Error: More actual than formal arguments in procedure call at (1)

Please could you investigate.

File: src/psyclone/domain/gocean/transformations/gocean_opencl_trans.py (hunk 4 above)
Comment (sergisiso) [reply src/psyclone/domain/gocean/transformations/gocean_opencl_trans.py:311]: Ups, of course, removed now.

**Comment by sergisiso** [comment]
@arporter This is ready for another look. The compilation issue was because the OpenCL transformation must be applied after the GOMoveBoundaries transformation and in that particular script that was not the case. I added a check in the OpenCL validation to get a proper transformation error message instead of keep going and have the compilation mismatch.

**Review by arporter — APPROVED** [review]
All requested changes have been made.
Tests and examples are all OK with compilation.
Docs all build OK.
This is 'towards' 1134 so I haven't checked for TODOs.

**Pull request merged**